arbitrator/prover/src/memory.rs

```diff
@@ -100,6 +139,11 @@ impl Memory {
 
     pub fn merkelize(&self) -> Cow<'_, Merkle> {
         if let Some(m) = &self.merkle {
+            for idx in self.dirty_indices.lock().unwrap().iter() {
+                let leaf_idx = idx / Self::LEAF_SIZE;
+                m.set(leaf_idx, hash_leaf(self.get_leaf_data(leaf_idx)));
+            }
+            self.dirty_indices.lock().unwrap().clear();
```

Comment:
I think there is a race condition here. You should lock the mutex only once, do the for loop, and then clear using the same lock.